Feat/GitHub CI/CD

LiilBoo opened the pull request:
1. minor changes to the CONTRIBUTING.md : explicitly and better explain how to run the test locally
2. ci.yaml : runs tests on ds_auth on every push to main
3. pr_approved.yaml : runs tests on ds_auth when a PR is approved
4. pr_merged.yaml : runs tests on ds_auth when PR is merged

- Notes -

 All github workflows can also be launched manually via the 'workflow dispatch' ;

All merge conflicts must be resolved in order to trigger a workflow

Choice is yours
you may change the triggers of the workflows
you may delete either workflows  if you don't want to run github actions when a PR is approved or merged
 

- Security -

Logically, the pr_approved and pr_merged tests should use the tests of disclone/main against the merged code (result of main code and PR code). translation: someone cannot modify the .github/workflows and have their own .github/workflows run as a result (which could be bad tests or broken tests or malicious code trying to extract secrets)


Pretty sure, I have done my research well, but let it be written that there's a difference between the events: pull_request and pull_request_target
Source: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

pull_request event should only have read-only access/rights with GITHUB_TOKEN 

pull_request_target ( not used in the workflows ) is an event with further rights (read/write/others) on the GITHUB_TOKEN

LiilBoo commented:
Also any test that doesn't pass, will mark the workflow as failed ( in red )

codewith-luke commented:
Can we see if we can do either:

1. A single "template workflow" that can be used for CI + pr_approved
2. Consolidate the conditions into a single workflow

Then we can drop the pr_merged.

Great stuff!

LiilBoo commented:
CHANGES:
- deleted pr_merged.yaml
- made the ci.yaml workflow callable by other workflows and re-usable
- made the pr_approved.yaml workflow call the ci.yaml (this reduces lines and make it dependent on ci.yaml )

1. refactored to the best of what I could find
2. After researching template-workflows, I could not find much documentation. They don't appeal to me as interesting. Maybe for people using Github Entreprise

Thanks for challenging me !
